bw_check_commit_hook/__init__.py

```diff
@@ -0,0 +1,205 @@
+        help="Path to config file",
+    )
+    return vars(parser.parse_args())
+
+
+def _load_config(config_file: Path) -> Optional[Dict]:
+    mandatory_keys = ["DIR_CONFIGS"]
+    try:
+        with config_file.open("r") as file:
+            config = yaml.safe_load(file)
+        for key in mandatory_keys:
+            if key not in config:
+                print(f"Missing mandatory key {key} in config file")
+                return None
+        return config
+    except Exception as e:
+        print(f"Error loading YAML config file: {e}")
+        return None
+
+
+def pre_commit_hook(running_in_ci: bool, commit_files: List[Path], config_file: Path) -> None:
+    config = _load_config(config_file)
+    if config is None:
+        sys.exit(1)
+
+    if running_in_ci:
+        commit_files = _get_commit_files_from_log()
+
+    errors = defaultdict(lambda: [])
+    # accumulate checkpatch errors for each configured dir
+    for config_dir, config in config["DIR_CONFIGS"].items():
+        # Default values
+        config['errors_enabled'] = config.get('errors_enabled', [])
+        config['errors_ignored'] = config.get('errors_ignored', [])
+        config['max_line_length'] = config.get('max_line_length', 120)
+
+        # Merge with defaults
+        config['errors_enabled'].extend(config.get("ERRORS_ENABLED", []))
+        config['errors_ignored'].extend(config.get("ERRORS_IGNORED", []))
+
+        patch_files = [
+            p
+            for p in commit_files
+            if _is_relative_to(p, config_dir)
+            and p not in config.get("IGNORED_FILES", [])
+            and not p.is_symlink()
+        ]
+        if patch_files:
+            commit_patch = _get_patch(running_in_ci, patch_files)
+            _run_checkpatch(commit_patch, errors, **config)
+
+    if errors:
+        sys.exit(1)
+
+def _get_commit_files_from_log() -> List[Path]:
+    cmd = ["git", "diff-tree", "--no-commit-id", "--name-only", "HEAD", "-r"]
+    log_output = check_output(cmd).decode().rstrip("\n")
+    return [Path(file_path) for file_path in log_output.split("\n")]
+
+
+def _get_patch(running_in_ci: bool, patch_files: List[Path]) -> str:
+    if running_in_ci:
+        return check_output(
+            [
+                "git",
+                "format-patch",
+                "--stdout",
+                "-1",
+                "HEAD",
+                "--",
+                *patch_files,
+            ]
+        ).decode()
+    commit_msg = _read_commit_msg_from_file()
+    return _diff_to_patch(patch_files, commit_msg)
+
+
+def _diff_to_patch(patch_files: List[Path], commit_msg: str) -> str:
+    diff = check_output(
+        ["git", "diff", "--cached", "--", *patch_files]
+    ).decode()
+    return PATCH_TEMPLATE.format(
+        diff=diff,
+        commit_msg=commit_msg,
+        current_datetime=datetime.now().strftime("%a, %d %b %Y %H:%M:%S %z"),
+    )
+
+
+def _is_relative_to(a_path: Path, another_path: Path) -> bool:
```

Comment:
It's deprecated in 3.12+ ttps://docs.python.org/3/library/pathlib.html#pathlib.PurePath.is_relative_to, so, better to use the abstraction to hide the version handling.

Reply:
The deprecation is for supporting multiple arguments, i.e. the signature will change from taking `*others` to taking `other` as a single argument. So it will still support our use in 3.12+ and can remain.